### ISSUE TITLE
Prevent cursor from moving after using an operator

### ISSUE THREAD

**Issue by seb-mueller** [neovim/neovim#12374]
- `nvim --version`: v0.4.3
- `vim -u DEFAULTS` (version: ) behaves differently? vim behaves the same (all version I'm aware of)
- Operating system/version: Ubuntu 19.10
- Terminal name/version: yakuake
- `$TERM`: xterm-256color

### Steps to reproduce using `nvim -u NORC`

```
nvim -u NORC
```
Copy the following paragraph in a buffer and position the cursor somewhere on line2:
```
line1
line2
```
Type `yip` to copy the whole paragraph in the unnamed register.

### Actual behaviour
cursor **moves** to line1

### Desired behaviour
cursor should **stay** in the same position it was typing `yip`

Above is only an example for yank. This behaviour is the same for other operators as well and it would be ideal to have an global option or flat to prevent the cursor from moving. 

### Further info
Having asked a [question](https://vi.stackexchange.com/questions/25503/keep-cursor-from-moving-when-using-operators/25507?noredirect=1#comment44904_25507) on stackexchange, it turns out this feature is not available in (n)vim. It's a behaviour that has annoyed me for quite some time.

**Comment by teto** [comment]
agreed that would be a nice fix. Patches welcome. I hope it won't break plugins too much.

**Comment by seb-mueller** [comment]
> I hope it won't break plugins too much.

True. Maybe it would be best to keep it as it is as default,  but introduce a flag such that it can be turned off optionally, i.e. `set nocursor_movement` or so.

**Comment by seb-mueller** [comment]
Just checking back to see if there is been any progress on this issue? It seems quite popular, so maybe a good enough reason to implement it?

**Comment by teto** [comment]
seems like it was not a good enough reason for people to devote time to this yet. You are welcome to try.

**Comment by seb-mueller** [comment]
Thanks, yes it seems like it, I'd love to try, but it's unfortunately beyond my programming skills (I'd probably need some hand holding), but I'll mention it in my next donation, maybe that bumps it up a little :)

**Comment by justinmk** [comment]
related 1789

**Comment by gh-liu** [comment]
You could use the `ModeChanged` event and the `vim.v.operator` variable to achieve this.

```lua
local api = vim.api
local autocmd = api.nvim_create_autocmd
local augroup = api.nvim_create_augroup

local g = augroup("user/keep_yank_position", { clear = true })

autocmd("ModeChanged", {
	pattern = { "n:no", "no:n" },
	group = g,
	callback = function(ev)
		if vim.v.operator == "y" then
			if ev.match == "n:no" then
				vim.b.user_yank_last_pos = vim.fn.getpos(".")
			else
				if vim.b.user_yank_last_pos then
					vim.fn.setpos(".", vim.b.user_yank_last_pos)
					vim.b.user_yank_last_pos = nil
				end
			end
		end
	end,
})

autocmd("ModeChanged", {
	pattern = {
		"V:n",
		"n:V",
		"v:n",
		"n:v",
	},
	group = g,
	callback = function(ev)
		local match = ev.match
		if vim.tbl_contains({ "n:V", "n:v" }, match) then
			-- vim.b.user_yank_last_pos = vim.fn.getpos(".")
			vim.b.user_yank_last_pos = vim.api.nvim_win_get_cursor(0)
		else
			-- if vim.tbl_contains({ "V:n", "v:n" }, match) then
			if vim.v.operator == "y" then
				local last_pos = vim.b.user_yank_last_pos
				if last_pos then
					-- vim.fn.setpos(".", last_pos)
					vim.api.nvim_win_set_cursor(0, last_pos)
				end
			end
			vim.b.user_yank_last_pos = nil
			-- end
		end
	end,
})

```

**Comment by telemachus** [comment]
@hinell Maybe combine gh-liu's tip with [this blog post](https://web.archive.org/web/20250327223946/https://nanotipsforvim.prose.sh/sticky-yank)? (NB: the original blog post is a 404, so this is an archive link.)

```lua
vim.keymap.set({ "n", "x" }, "y", function()
    vim.b.cursorPreYank = vim.api.nvim_win_get_cursor(0)
    return "y"
end, { expr = true })
vim.keymap.set("n", "Y", function()
    vim.b.cursorPreYank = vim.api.nvim_win_get_cursor(0)
    return "y$"
end, { expr = true })

vim.api.nvim_create_autocmd("TextYankPost", {
    callback = function()
        if vim.v.event.operator == "y" and vim.b.cursorPreYank then
            vim.api.nvim_win_set_cursor(0, vim.b.cursorPreYank)
            vim.b.cursor_pre_yank = nil
        end
    end,
})
```

I've been using this code for a while, and it works well. YMMV

**Comment by hinell** [comment]
@telemachus Thanks. Even though I would prefer this being implemented by nvim I've actually settled down with `CursorMoved` and `TextYankPost` for now. Works pretty good. No need to rebind keymaps. Going to make it into a small plugin.

My concern though is that significant perf downgrade may happen due to `CursorMoved` being constantly called.

May be it's time to explore alternatives with `getchar()` ...

**Comment by llakala** [comment]
> [@hinell](https://github.com/hinell) Maybe combine gh-liu's tip with [this blog post](https://nanotipsforvim.prose.sh/sticky-yank)?

Dead link - here's an [archive](https://web.archive.org/web/20250121072908/https://nanotipsforvim.prose.sh/sticky-yank).

**Comment by telemachus** [comment]
> Dead link - here's an [archive](https://web.archive.org/web/20250121072908/https://nanotipsforvim.prose.sh/sticky-yank).

@llakala Thanks. The author of that site seems to have taken it offline, but you can still see [the latest version of the suggestion in the archived GitHub repo](https://github.com/chrisgrieser/nanotipsforvim-blog/blob/main/articles/sticky-yank.md). I updated my answer with an archive link. Thanks for the nudge.

**Comment by llakala** [comment]
I tried using the new `CmdAtom` to achieve this in a more generalized way than `TextYankPost`:
```nix
vim.api.nvim_create_autocmd("CmdAtom", {
  callback = function(ev)
    if ev.data.type == "operator" and ev.data.operator == "y" and ev.data.moved then
      vim.api.nvim_win_set_cursor(0, ev.data.pos)
    end
  end,
})
```
However, this seems to flash the cursor position _sometimes_, when an update happens between the yank and CmdAtom being triggered. I'm not sure if there's a way to avoid this - maybe the CmdAtom event should be emitted earlier somehow?

**Comment by justinmk** [comment]
@llakala does it help if you `vim.schedule()` the nvim_win_set_cursor call?

WDYM by "flash the cursor position" ? Is that a cosmetic issue or a functionality issue?

**Comment by llakala** [comment]
> does it help if you vim.schedule() the nvim_win_set_cursor call?

No, still rubber-bands.

> WDYM by "flash the cursor position" ? Is that a cosmetic issue or a functionality issue?

Cosmetic - if you do `yk`, you can see the cursor move to the line above, and then immediately move _back_ to the original position.
